src/Admin/VMware.ScriptRuntimeService.AdminEngine/VCRegistration/VCTrustedCertificatesCollector.cs

```diff
@@ -0,0 +1,275 @@
+                  endIndex = cert_chain.IndexOf(PEM_END, startIndex);
+               }
+            }
+
+            return result;
+         }
+      }
+
+      private async Task<IEnumerable<string>> GetTrustedRootChainsChainsAsync(HttpClient client) {
+         using (var response = await client.GetAsync("/api/vcenter/certificate-management/vcenter/trusted-root-chains")) {
+            response.EnsureSuccessStatusCode();
+
+            var responseStr = await response.Content.ReadAsStringAsync();
+            dynamic result = JsonConvert.DeserializeObject<dynamic>(responseStr);
+
+            if (result is IEnumerable enumerable) {
+               return enumerable.Cast<dynamic>().Where(r => r?.chain != null).Select(c => (string) c.chain);
+            } else {
+               throw new TrustedCertificateRetrievalException("/api/vcenter/certificate-management/vcenter/trusted-root-chains did not return an array");
+            }
+         } 
+      }
+
+      private HttpClient GetHttpClient() {
+         _logger.LogInformation("Creating http client");
+
+         HttpMessageHandler clientHandler = new HttpClientHandler() {
+            ServerCertificateCustomValidationCallback = (message, certificate, chain, policyErrors) => {
+
+               if (certificate == null) { throw new ArgumentNullException(nameof(certificate)); }
+
+               if (policyErrors == System.Net.Security.SslPolicyErrors.None) {
+                  return true;
+               } else if (_ignoreServerCertificateValidation) {
+                  return true;
+               } else {
+                  var normalizedThumbprint = _thumbprint?.Replace(":", "")?.Replace(" ", "")?.ToUpperInvariant();
+                  var sha1Thumbprint = certificate.GetCertHashString(HashAlgorithmName.SHA1);
+                  var sha256Thumbprint = certificate.GetCertHashString(HashAlgorithmName.SHA256);
+
+                  if (!sha1Thumbprint.Equals(normalizedThumbprint) && !sha256Thumbprint.Equals(normalizedThumbprint)) {
+                     _logger.LogError(
+                        $"VC [${_psc}] certificate thumbprint [sha1:{sha1Thumbprint}, sha256:{sha256Thumbprint}] does not match the supplied one [{_thumbprint}]");
+                  }
+
+                  return sha1Thumbprint.Equals(normalizedThumbprint) || sha256Thumbprint.Equals(normalizedThumbprint);
+               }
+            }
+         };
+
+         if (_httpMessageHandlerFactory != null) {
+            clientHandler = _httpMessageHandlerFactory.Invoke(new LoggingHandler(_loggerFactory, clientHandler));
+         }
+
+         return new HttpClient(clientHandler) {
+            BaseAddress = new Uri($"https://{_psc}")
+         };
+      }
+
+      private async Task AuthenticateClientAsync(HttpClient client) {
+         _logger.LogInformation("Establishing session");
+
+         string authHeaderValue = Convert.ToBase64String(Encoding.ASCII.GetBytes(_username + ":" + SecureStringToString(_password)));
```

Comment:
Would this value end up in the log? If so, can you change the logging to not include the value of this type of headers?

Reply:
Yes, it will. I'll find a way to skip the Auth header